osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs

```diff
@@ -0,0 +1,183 @@
+
+        // rate for a hit is calculated using the end time of another hit object earlier in time
+        // caching them here for easy access
+        private readonly Dictionary<HitObject, double> previousEndTimes = new Dictionary<HitObject, double>();
+
+        // record the value removed from recentRates when an object is hit
+        // for rewind support
+        private readonly Dictionary<HitObject, double> dequeuedRates = new Dictionary<HitObject, double>();
+
+        public ModAdaptiveSpeed()
+        {
+            InitialRate.BindValueChanged(val => SpeedChange.Value = val.NewValue);
+            AdjustPitch.BindValueChanged(applyPitchAdjustment);
+        }
+
+        public void ApplyToHUD(HUDOverlay overlay)
+        {
+            // this is only used to transform the SpeedChange bindable
+            this.overlay = overlay;
+        }
+
+        public void ApplyToTrack(ITrack track)
+        {
+            this.track = track;
+
+            InitialRate.TriggerChange();
+            AdjustPitch.TriggerChange();
+            recentRates.Clear();
+            recentRates.AddRange(Enumerable.Range(0, average_count).Select(_ => InitialRate.Value));
+        }
+
+        public void ApplyToSample(DrawableSample sample)
+        {
+            sample.AddAdjustment(AdjustableProperty.Frequency, SpeedChange);
+        }
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
+
+        private void applyPitchAdjustment(ValueChangedEvent<bool> adjustPitchSetting)
+        {
+            // remove existing old adjustment
+            track?.RemoveAdjustment(adjustmentForPitchSetting(adjustPitchSetting.OldValue), SpeedChange);
+
+            track?.AddAdjustment(adjustmentForPitchSetting(adjustPitchSetting.NewValue), SpeedChange);
+        }
+
+        private AdjustableProperty adjustmentForPitchSetting(bool adjustPitchSettingValue)
+            => adjustPitchSettingValue ? AdjustableProperty.Frequency : AdjustableProperty.Tempo;
+
+        public void ApplyToDrawableHitObject(DrawableHitObject drawable)
+        {
+            drawable.OnNewResult += (o, result) =>
+            {
+                if (dequeuedRates.ContainsKey(result.HitObject)) return;
+
+                if (!result.IsHit) return;
+                if (!result.Type.AffectsAccuracy()) return;
+                if (!previousEndTimes.ContainsKey(result.HitObject)) return;
+
+                double prevEndTime = previousEndTimes[result.HitObject];
+
+                recentRates.Add(Math.Clamp((result.HitObject.GetEndTime() - prevEndTime) / (result.TimeAbsolute - prevEndTime) * SpeedChange.Value, slowest_rate, fastest_rate));
+
+                dequeuedRates.Add(result.HitObject, recentRates[0]);
+                recentRates.RemoveAt(0);
+
+                overlay.TransformBindableTo(SpeedChange, recentRates.Average(), 100);
```

Comment:
This is going to create a lot of transforms, potentially overlapping, which is going to be a performance concern. The usual way of handling that is using `Interpolation.DampContinuously()` or similar.

Reply:
Thanks, that's the function I was missing. Wasn't sure about using transforms in the first place.